go/new_merkletree/merkle_tree.go

```diff
@@ -0,0 +1,177 @@
+package new_merkletree
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"hash"
+	"math"
+)
+
+// New creates a new merkle hash tree.  The number of "leaves" of the tree,
+// as well as their contents, are retrieved through |dao|.  If you wish to
+// have acceptable performance on non-trivial tree sizes, you'll want to
+// provide |cache| (otherwise, pass `nil`).  The hash function used for all
+// nodes in the tree is specified by |hasher|.
+func New(dao NewMerkleTreeDataInterface, cache NewMerkleTreeCacheInterface, hasher hash.Hash) *NewMerkleTree {
+	return &NewMerkleTree{dao: dao, cache: cache, hasher: hasher}
+}
+
+func (mt NewMerkleTree) CurrentRoot() (Hash, error) {
+	if mt.dao.Size() == 0 {
+		// Special case: empty trees get hashes of empty strings
+		mt.hasher.Reset()
+		return mt.hasher.Sum([]byte{}), nil
+	} else {
```

Comment:
No need for an `else` block here, just have the next return directly below the `if` block.